File: its-my-training-day/leetcode-top-75/questions/matrix/matrix-in-place/set-matrix-zeroes/solution.py
import os
import sys
from typing import List, Dict, Optional, Tuple

ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../"))
if ROOT_DIR not in sys.path:
    sys.path.append(ROOT_DIR)

from shared.python.ds import ArrayList, ListNode, Stack, Queue, TreeNode, MinHeap, MaxHeap

def set_matrix_zeroes(matrix: List[List[int]]) -> None:
    rows = len(matrix)
    cols = len(matrix[0])
    row_zero = False
    col_zero = False
    r = 0
    while r < rows:
        if matrix[r][0] == 0:
            col_zero = True
        r += 1
    c = 0
    while c < cols:
        if matrix[0][c] == 0:
            row_zero = True
        c += 1
    r = 1
    while r < rows:
        c = 1
        while c < cols:
            if matrix[r][c] == 0:
                matrix[r][0] = 0
                matrix[0][c] = 0
            c += 1
        r += 1
    r = 1
    while r < rows:
        if matrix[r][0] == 0:
            c = 1
            while c < cols:
                matrix[r][c] = 0
                c += 1
        r += 1
    c = 1
    while c < cols:
        if matrix[0][c] == 0:
            r = 1
            while r < rows:
                matrix[r][c] = 0
                r += 1
        c += 1
    if row_zero:
        c = 0
        while c < cols:
            matrix[0][c] = 0
            c += 1
    if col_zero:
        r = 0
        while r < rows:
            matrix[r][0] = 0
            r += 1

File: its-my-training-day/leetcode-top-75/questions/matrix/matrix-in-place/set-matrix-zeroes/solution.cpp
#include "../../../../shared/cpp/ds.hpp"

#include <algorithm>
#include <cctype>
#include <cstdint>
#include <functional>
#include <string>
#include <unordered_map>
#include <vector>

void setMatrixZeroes(std::vector<std::vector<int>>& matrix) {
    int rows = static_cast<int>(matrix.size());
    int cols = static_cast<int>(matrix[0].size());
    bool rowZero = false;
    bool colZero = false;
    for (int r = 0; r < rows; ++r) {
        if (matrix[r][0] == 0) colZero = true;
    }
    for (int c = 0; c < cols; ++c) {
        if (matrix[0][c] == 0) rowZero = true;
    }
    for (int r = 1; r < rows; ++r) {
        for (int c = 1; c < cols; ++c) {
            if (matrix[r][c] == 0) {
                matrix[r][0] = 0;
                matrix[0][c] = 0;
            }
        }
    }
    for (int r = 1; r < rows; ++r) {
        if (matrix[r][0] == 0) {
            for (int c = 1; c < cols; ++c) {
                matrix[r][c] = 0;
            }
        }
    }
    for (int c = 1; c < cols; ++c) {
        if (matrix[0][c] == 0) {
            for (int r = 1; r < rows; ++r) {
                matrix[r][c] = 0;
            }
        }
    }
    if (rowZero) {
        for (int c = 0; c < cols; ++c) matrix[0][c] = 0;
    }
    if (colZero) {
        for (int r = 0; r < rows; ++r) matrix[r][0] = 0;
    }
}
